src/workerd/io/worker.c++

```diff
@@ -2316,8 +2300,237 @@ public:
+      outgoingQueueNotifier->notify();
+    }
+
+  private:
+    struct MessageQueue {
+      kj::Vector<kj::String> messages;
+      size_t head;
+      enum class Status { ACTIVE, PAUSED, CLOSED } status;
+    };
+
+    static kj::Maybe<kj::String> pollMessage(MessageQueue& messageQueue) {
+      if (messageQueue.head < messageQueue.messages.size()) {
+        kj::String message = kj::mv(messageQueue.messages[messageQueue.head++]);
+        if (messageQueue.head == messageQueue.messages.size()) {
+          messageQueue.head = 0;
+          messageQueue.messages.clear();
+        }
+        return kj::mv(message);
+      }
+      return {};
+    }
+
+    static void deliverProtocolMessages(kj::Maybe<InspectorChannelImpl&>& channel,
+                                        kj::Locked<MessageQueue>& incomingQueue) {
+      KJ_REQUIRE(incomingQueue->status == MessageQueue::Status::ACTIVE);
+      KJ_IF_MAYBE(c, channel) {
+        kj::ArrayPtr<kj::String> messages = incomingQueue->messages.slice(
+            incomingQueue->head, incomingQueue->messages.size());
+        c->dispatchProtocolMessages(messages);
+        incomingQueue->messages.clear();
+        incomingQueue->head = 0;
+      }
+    }
+
+    void shutdown() {
+      // Wake any waiters.
+      outgoingQueueNotifier->notify();
+
+      // Drain incoming queue, the isolate thread may be waiting on it
+      // on will notice it is closed if woken without any messages to
+      // deliver in WebSocketIoWorker::waitForMessage().
+      {
+        auto lockedIncomingQueue = incomingQueue.lockExclusive();
+        lockedIncomingQueue->head = 0;
+        lockedIncomingQueue->messages.clear();
+        lockedIncomingQueue->status = MessageQueue::Status::CLOSED;
+      }
+      {
+        auto lockedOutgoingQueue = outgoingQueue.lockExclusive();
+        lockedOutgoingQueue->status = MessageQueue::Status::CLOSED;
+      }
+    }
+
+    kj::Promise<void> incomingLoop() {
+      for (;;) {
+        auto message = co_await webSocket.receive();
+        KJ_SWITCH_ONEOF(message) {
+          KJ_CASE_ONEOF(text, kj::String) {
+            auto lockedIncomingQueue = incomingQueue.lockExclusive();
+            lockedIncomingQueue->messages.add(kj::mv(text));
+            if (lockedIncomingQueue->status == MessageQueue::Status::ACTIVE) {
+              deliverProtocolMessages(channel, lockedIncomingQueue);
+            }
+          }
+          KJ_CASE_ONEOF(blob, kj::Array<byte>){
+            // Ignore.
+          }
+          KJ_CASE_ONEOF(close, kj::WebSocket::Close) {
+            shutdown();
+          }
+        }
+      }
+    }
+
+    kj::Promise<void> outgoingLoop() {
+      for (;;) {
+        co_await outgoingQueueNotifier->awaitNotification();
+        try {
+          auto lockedOutgoingQueue = outgoingQueue.lockExclusive();
+          if (lockedOutgoingQueue->status == MessageQueue::Status::CLOSED) {
+            co_await webSocket.close(1000, "client closed connection");
```

Comment:
It looks like the behavior changed slightly in the coroutine conversion, previously we would send any remaining messages before closing.

Reply:
Ack, fixed, thanks.